### PR TITLE
matchers: contain all

### PR DESCRIPTION
**Pull request by MykolaGolubyev**

### CODE REVIEW

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Test
+    void "should pass when value contains all expected values"() {
+        actual(['a', 'b', 'd']).should(containAll('b', 'a'))
+    }
+
+    @Test
+    void "should throw exception when value contain expected value, but should not"() {
+        code {
+            actual(['a', 'b', 'd']).shouldNot(containAll('b', 'a'))
```
Comment (MykolaGolubyev) [inline webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:50]: @tsiq-clemens this one is a bit tricky (at least right now evening time). Should it fail if at least one is contained or only if all are contained?

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java [hunk 2]
```diff
@@ -89,6 +108,54 @@ public static ContainMatcher containing(Object expected) {
...
+     * @return matcher instance
+     */
+    public static ContainAllMatcher containAll(Object... expected) {
+        return new ContainAllMatcher(Arrays.asList(expected));
+    }
+
+    /**
+     * Containing all matcher. Alias to containAll
+     * <pre>
+     * actual(collection).should(containAll(list));
```
Comment (tsiq-clemens) [inline webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java:138]: more like `myArray.should == [a, b, containingAll(list)]` , here perhaps translated to java

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java [hunk 3]
```diff
@@ -89,6 +108,54 @@ public static ContainMatcher containing(Object expected) {
...
+     * @return matcher instance
+     */
+    public static ContainAllMatcher containingAll(Collection<Object> expected) {
+        return new ContainAllMatcher(expected);
+    }
+
+    /**
+     * Containing all matcher. Alias to containAll
+     * <pre>
+     * actual(collection).should(containAll(2, 3, "a"));
```
Comment (tsiq-clemens) [inline webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java:150]: similar example comment

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcher.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @Override
+    public String negativeMismatchedMessage(ActualPath actualPath, Object actual) {
+        return actualPath + " expect to not contain all " + DataRenderers.render(expectedList) + "\n" +
+                containAnalyzer.generateMismatchReport();
+    }
+
+    @Override
+    public boolean negativeMatches(ActualPath actualPath, Object actual) {
```
Comment (tsiq-clemens) [inline webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcher.java:81]: is this not simply the negative of `matches`?

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+package org.testingisdocumenting.webtau.expectation.contain
+
+import org.junit.Test
+
+import static org.testingisdocumenting.webtau.Matchers.*
+
+class ContainAllMatcherTest {
+    @Test
+    void "should throw exception when value doesn't contain all expected value"() {
```
Comment (tsiq-clemens) [inline webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:25]: typo: values

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcher.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @Override
+    public String matchedMessage(ActualPath actualPath, Object actual) {
+        return "contains all " + DataRenderers.render(expectedList);
+    }
+
+    @Override
+    public String mismatchedMessage(ActualPath actualPath, Object actual) {
+        return actualPath + " expect to contain all " + DataRenderers.render(expectedList) + "\n" +
```
Comment (tsiq-clemens) [inline webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcher.java:48]: "expects to contain all" ?

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                '                                ^')
+    }
+
+    @Test
+    void "should pass when value contains all expected values"() {
+        actual(['a', 'b', 'd']).should(containAll('b', 'a'))
+    }
+
+    @Test
+    void "should throw exception when value contain all expected values, but should not"() {
```
Comment (tsiq-clemens) [inline webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:48]: typo: "contains all"

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy (hunk 1 above)
Comment (tsiq-clemens) [reply webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:50]: it should fail iff `should(containAll(list))` succeeds. 
So if actual contains some but not all of list, that's a success

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        code {
+            actual(['a', 'b', 'd']).shouldNot(containAll('b', 'a'))
+        } should throwException('\n[value] expect to not contain all [b, a]\n' +
+                '[value][1]: equals "b"\n' +
+                '[value][0]: equals "a"')
+    }
+
+    @Test
+    void "should pass when value contain only some expected values, but should not"() {
```
Comment (tsiq-clemens) [inline webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:57]: "contains only some"

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        actual(['a', 'b', 'd']).shouldNot(containAll('b', 'a', 'x'))
+    }
+
+    @Test
+    void "should pass when value not contains all expected values"() {
```
Comment (tsiq-clemens) [inline webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:62]: These descriptions get confusing for `shouldNot` tests :-)
How about phrasing them something like "negative matcher should pass if value contains none of the expected values"

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+ * limitations under the License.
+ */
+
+package org.testingisdocumenting.webtau.expectation.contain
+
+import org.junit.Test
+
+import static org.testingisdocumenting.webtau.Matchers.*
+
+class ContainAllMatcherTest {
```
Comment (tsiq-clemens) [inline webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:23]: add tests for the alias?

File: webtau-http-groovy/src/test/groovy/org/testingisdocumenting/webtau/http/HttpGroovyTest.groovy [hunk 11]
```diff
@@ -539,6 +539,30 @@ class HttpGroovyTest extends HttpTestBase {
         body.get("list[2]").traceableValue.checkLevel.should == CheckLevel.None
     }
 
+    @Test
```
Comment (tsiq-clemens) [inline webtau-http-groovy/src/test/groovy/org/testingisdocumenting/webtau/http/HttpGroovyTest.groovy:542]: add an example for the alias (now that you put work into generating it)?

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy (hunk 1 above)
Comment (MykolaGolubyev) [reply webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:50]: yeah, makes sense, I updated code after this comment

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java (hunk 2 above)
Comment (MykolaGolubyev) [reply webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java:138]: this makes sense, thank you

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcher.java (hunk 6 above)
Comment (MykolaGolubyev) [reply webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcher.java:48]: right! thanks

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy (hunk 10 above)
Comment (MykolaGolubyev) [reply webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:23]: makes sense

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy (hunk 9 above)
Comment (MykolaGolubyev) [reply webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:62]: this is very good, I will re-phrase

File: webtau-http-groovy/src/test/groovy/org/testingisdocumenting/webtau/http/HttpGroovyTest.groovy (hunk 11 above)
Comment (MykolaGolubyev) [reply webtau-http-groovy/src/test/groovy/org/testingisdocumenting/webtau/http/HttpGroovyTest.groovy:542]: what do you mean by generating it?

**Review by MykolaGolubyev — COMMENTED** [review]
Thank you for the review. Will update soon.

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java [hunk 12]
```diff
@@ -89,6 +108,54 @@ public static ContainMatcher containing(Object expected) {
...
+     * @return matcher instance
+     */
+    public static ContainAllMatcher containAll(Object... expected) {
+        return new ContainAllMatcher(Arrays.asList(expected));
+    }
+
+    /**
+     * Containing all matcher. Alias to containAll
+     * <pre>
+     * actual(listOfLists).should(contain(containAll(myList)));
```
Comment (tsiq-clemens) [inline webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java:138]: I'd expect this javadoc example to use `containingAll`

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java [hunk 13]
```diff
@@ -89,6 +108,54 @@ public static ContainMatcher containing(Object expected) {
...
+     * @return matcher instance
+     */
+    public static ContainAllMatcher containingAll(Collection<Object> expected) {
+        return new ContainAllMatcher(expected);
+    }
+
+    /**
+     * Containing all matcher. Alias to containAll
+     * <pre>
+     * actual(listOfLists).should(contain(containAll(2, 3, "a")));
```
Comment (tsiq-clemens) [inline webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java:150]: same comment: use `containingAll`

File: webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy [hunk 14]
```diff
@@ -40,26 +40,31 @@ class ContainAllMatcherTest {
...
 
     @Test
-    void "should pass when value contains all expected values"() {
+    void "passes when all values are present"() {
         actual(['a', 'b', 'd']).should(containAll('b', 'a'))
     }
 
     @Test
-    void "should throw exception when value contain all expected values, but should not"() {
+    void "negative matcher fails only when all the values are present "() {
```
Comment (tsiq-clemens) [inline webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/contain/ContainAllMatcherTest.groovy:48]: love it

File: webtau-http/src/test/resources/prices.json [hunk 15]
```diff
@@ -0,0 +1,10 @@
+[
+  {
+    "id": "game1",
+    "prices": [15, 30, 40, 60]
+  },
+  {
+    "id": "game2",
+    "prices": [10, 30]
+  }
+]
```
Comment (tsiq-clemens) [inline webtau-http/src/test/resources/prices.json:10]: not sure if you care about missing newlines at the end of a file

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java (hunk 12 above)
Comment (MykolaGolubyev) [reply webtau-core/src/main/java/org/testingisdocumenting/webtau/Matchers.java:138]: ah I thought I actually typed containing... thanks

File: webtau-http/src/test/resources/prices.json (hunk 15 above)
Comment (MykolaGolubyev) [reply webtau-http/src/test/resources/prices.json:10]: not really :), adding one

**Pull request merged**